tls1.3 cipher suites support

deweerdt opened the pull request:
This commit adds a `cipher-suite-tls1.3` configuration item, lists the
cipher suites specific to TLSv1.3.
Currently supported values are:
- `TLS13-AES-256-GCM-SHA384`
- `TLS13-CHACHA20-POLY1305-SHA256`
- `TLS13-AES-128-GCM-SHA256`

Default value is:
`TLS13-AES-256-GCM-SHA384:TLS13-CHACHA20-POLY1305-SHA256:TLS13-AES-128-GCM-SHA256`

Notes: this will error out if `TLS_AES_128_GCM_SHA256` is left out the
list, since it's support is mandatory for TLS1.3 implementations.

TODO:
- [ ] Add docs

omasanori commented:
Great improvement!

However, I doubt if the conformance check is relevant.
As the previous versions of TLS, the requirement is limited by the phrase "In the absence of an application profile standard specifying otherwise" and the existing code does not force mandatory cipher suite for older versions i.e. TLS_RSA_WITH_AES_128_CBC_SHA.
 (Yes, unlike TLS 1.3, we already have good reasons why we do not do there: no FS, no AEAD, no EC, ...)

Interoperability warning would be useful anyway.

deweerdt commented:
Thanks for the suggestion, @omasanori . 4951ae6 makes the absence of `TLS_AES_128_GCM_SHA256` a   warning.

omasanori commented:
Thank you!

Ys88 commented:
Hi,

Any idea when this commit will be applied to the Master branch? Thank you.

kazuho commented on src/main.c:
```diff
@@ -726,6 +770,13 @@ static int listener_setup_ssl(h2o_configurator_command_t *cmd, h2o_configurator_
+        ERR_print_errors_cb(on_openssl_print_errors, stderr);
+        goto Error;
+    }
+#endif
```
I think we do not need this, because we consistently use picotls for handling TLS 1.3 connections.

kazuho commented on src/main.c:
```diff
@@ -479,11 +514,20 @@ static const char *listener_setup_ssl_picotls(struct listener_config_t *listener
+    }
+#else
+    ptls_cipher_suite_t **cipher_suites = ptls_openssl_cipher_suites;
+#endif
```
I think we need to handle the arguments of `cipher-suites-tls1.3` directly for picotls. Otherwise, users would not be able to customize the cipher-suites when using OpenSSL older than 1.1.0.

deweerdt replied:
Would you move the parsing of the cipher description strings to picotls, or keep them in `src/main.c` in h2o?

kazuho replied:
I think we can keep it outside of picotls.

It's the application's responsibility to build the cipher-suites array (`ptls_context_t::cipher_suites`), including allocating sufficient memory for holding the list. I do not think there is a good way of providing the parsing function from inside of picotls considering these requirements.

OscarWater commented:
Hello,

I am new here, so apologize if this is not the right place to post. 

I was wondering if the above has been merged into the master branch of H2O ? Because I am enable to force the use of TLS13-CHACHA20-POLY1305-SHA256 as the primary cipher method.

Thank you !

Ys88 commented:
Hi,

Sorry to bother you. I know it's not very important, but is there a chance this be merged into the master in a near future?

Thank you.

deweerdt commented:
@kazuho sorry about the long hiatus here. I've addressed all your comments so far i believe. This PR includes the changes from https://github.com/h2o/picotls/pull/335 to allow server side order selection for picotls.

deweerdt commented:
The openssl from the docker test image doesn't know about `-tls1_3`, should i update it?

kazuho commented on src/main.c:
```diff
@@ -730,19 +733,66 @@ static h2o_iovec_t *build_http2_origin_frame(h2o_configurator_command_t *cmd, yo
     return http2_origin_frame;
 }
 
+static ptls_cipher_suite_t **parse_tls13_ciphers(h2o_configurator_command_t *cmd, yoml_t *node, const char *tls13_ciphers)
```
I do not think we need to pass `tls13_ciphers` as a separate argument, as it can be obtained as `node->data.scalar`?

kazuho commented on src/main.c:
```diff
@@ -730,19 +733,66 @@ static h2o_iovec_t *build_http2_origin_frame(h2o_configurator_command_t *cmd, yo
+{
+    char *p = alloca(strlen(tls13_ciphers) + 1), *saveptr = NULL;
+    strcpy(p, tls13_ciphers);
+    struct {
```
```suggestion
    static const struct {
```

kazuho commented on src/main.c:
```diff
@@ -730,19 +733,66 @@ static h2o_iovec_t *build_http2_origin_frame(h2o_configurator_command_t *cmd, yo
+        const char *name;
+        ptls_cipher_suite_t *cipher;
+    } cipher_suites[] = {
+        { "TLS_AES_128_GCM_SHA256", &ptls_openssl_aes128gcmsha256},
```
How about making the following changes:
* Instead of using `const char *` to hold the names, use `h2o_iovec_t`. The names can be stored like `{H2O_STRLIT("TLS_AES_128_GCM_SHA256")}`.
* Instead of creating a copied string, use `strchr(const_str, ':')` to split the list of cipher-suites being named, then use `h2o_memis` to compare them.

kazuho commented on src/main.c:
```diff
@@ -730,19 +733,66 @@ static h2o_iovec_t *build_http2_origin_frame(h2o_configurator_command_t *cmd, yo
+        int found  = 0;
+        for (int i = 0; i < (sizeof(cipher_suites) / sizeof(cipher_suites[0])); i++) {
+            if (!strcmp(p, cipher_suites[i].name)) {
+                ret.entries = h2o_mem_realloc(ret.entries, sizeof(ret.entries[0]) * (ret.size + 1));
```
I think we'd prefer using `h2o_vector_reserve`, as `ret` is an H2O_VECTOR.

kazuho commented on src/main.c:
```diff
@@ -730,19 +733,66 @@ static h2o_iovec_t *build_http2_origin_frame(h2o_configurator_command_t *cmd, yo
+        }
+        p = strtok_r(NULL, ":", &saveptr);
+    }
+    ret.entries = h2o_mem_realloc(ret.entries, sizeof(ret.entries[0])  * (ret.size + 1));
```
Please see above.

kazuho commented on src/main.c:
```diff
@@ -730,19 +733,66 @@ static h2o_iovec_t *build_http2_origin_frame(h2o_configurator_command_t *cmd, yo
+    ret.entries[ret.size++] = NULL;
+
+    if (!seen_tls_aes_128_gcm_sha256) {
+        h2o_configurator_errprintf(cmd, node, "Warning: not enabling TLS_AES_128_GCM_SHA256 might reduce TLS1.3 interoperability, see RFC 8446 9.1");
```
:+1:

kazuho commented on src/main.c:
```diff
@@ -730,19 +733,66 @@ static h2o_iovec_t *build_http2_origin_frame(h2o_configurator_command_t *cmd, yo
     uint64_t ocsp_update_interval = 4 * 60 * 60; /* defaults to 4 hours */
     unsigned ocsp_max_failures = 3;              /* defaults to 3; permit 3 failures before temporary disabling OCSP stapling */
     int use_neverbleed = 1, use_picotls = 1;     /* enabled by default */
+    ptls_cipher_suite_t **tls13_cipher_suites = NULL;
```
The naming convention that we have here is:
* the variable that holds the converted value uses the same name as the config parameter
* the variable that holds the `yoml_t *` has the `_node` suffix.

Following that convention, I think we want to name this `cipher_suite_tls13` and `yoml_t *` above as `cipher_suite_tls13_node`.

Separately, I think it might make sense to set the default to `ptls_openssl_cipher_suites` here, and in `listener_setup_ssl_picotls` always use the supplied value.

kazuho reviewed (commented):
Thank you for the PR. I appreciate the tests.

The changes Look good to me, my comments below are all editorial, I believe.

deweerdt commented:
closing in favor or https://github.com/h2o/h2o/pull/2653
